### PR TITLE
[FLINK-4669] scala api createLocalEnvironment() function add default Configuration parameter

### PR DESCRIPTION
**Pull request by shijinkui**
add `createLocalEnvWithWebUI` API in env.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-4669] scala api createLocalEnvironment() function add default Configuration parameter")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala [hunk 1]
```diff
@@ -124,7 +125,7 @@ class StreamExecutionEnvironment(javaEnv: JavaEnv) {
    * Gets the checkpoint config, which defines values like checkpoint interval, delay between
    * checkpoints, etc.
    */
-  def getCheckpointConfig = javaEnv.getCheckpointConfig()
```
Comment (StephanEwen) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala]: Calling Java methods with parenthesis or without actually somewhat subject to debate.

For Scala methods, the writer decides whether it is a method with or without parenthesis.
For Java, the caller can only "guess" what the method does (is returning a mutable field considered a side effect?), so in most places we call Java methods as Java methods.

Comment (shijinkui) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala]: IDEA ful-tip:
Methods that follow JavaBean naming contract for accessors are expected to have no side effects. The recommended convention is to use a parameterless method whenever there are no parameters and the method have no side effect. This convention supports the uniform access principle, which says that client code should not be affected by a decision to implement an attribute as a field or method. The problem is that Java does not implement the uniform access principle. To bridge that gap, Scala allows you to leave off the empty parentheses on an invocation of function that takes no arguments. * Refer to Programming in Scala, 10.3 Defining parameterless methods

Comment (shijinkui) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala]: http://stackoverflow.com/questions/8124055/scala-tostring-parenthesize-or-not   http://stackoverflow.com/questions/12334936/why-does-scala-need-parameterless-in-addition-to-zero-parameter-methods

Comment (StephanEwen) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala]: That is fine, IDEA may have different recommendations, Flink can define its own guidelines.
I (and others) don't agree with everything that the Scala folks recommend.
And even in that description, it is phrased as a possibility to bridge the gap between Scala's and Java's features.

Comment (shijinkui) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala]: That's fine.

**Comment by StephanEwen** [comment]
Code looks good, but the build detected a break in binary compatibility. Have a look at the Travis logs.

Changing the `createLocalEnvironment` to use an additional parameter with default value broke the compatibility. The fix would be to have an extra method (and no default parameter value), not change the existing method.

**Comment by shijinkui** [comment]
@StephanEwen  I‘ve a new function name `createCustomLocalEnv`. Sorry for late ack.

**Comment by StephanEwen** [comment]
Okay, this has binary compatibility now.
Few things I would still suggest to do:
  - There is still a bug that the `parallelism` and `port` parameter are mixed up.
  - It still does not use the DEFAULT_LOCAL_PARALLELISM.
  - The same methods should exist in all environments (batch / streaming / Java / Scala)

**Comment by shijinkui** [comment]
hi, @StephanEwen 
follow your exact suggest, have add the functions at four place.

**Comment by shijinkui** [comment]
@StephanEwen  have any improvement needed?

**Comment by StephanEwen** [comment]
Sorry for asking for another iteration, but here are some comments:

  - The other Java code uses the primitive `int` everywhere, rather than the boxed integer. Let's follow that style.

  - I would omit the `public static ExecutionEnvironment createCustomLocalEnv(Integer parallelism, Configuration config)` variant. The benefit is very small compared to the `public static ExecutionEnvironment createLocalEnvironment(Configuration config)` method, but it blows up the API into too many variants.

  - Similarly, I would only have the `public static ExecutionEnvironment createLocalEnvWithWebUI (int port, Configuration config)` and drop the variant with the additional `parallelism` parameter.

In general, I think it is good to not have too many method options, but only those that add significant functionality. For the method that starts the WebUI, that is the case, for the method with the additional parallelism, probably not.

**Comment by shijinkui** [comment]
> In general, I think it is good to not have too many method options, but only those that add significant functionality. For the method that starts the WebUI, that is the case, for the method with the additional parallelism, probably not.

Function having least parameters is the best style. If have no enough parameter for user, the function's ability is limited. That's why I prefer Scala, as it has default parameter value, keep the both.

> The other Java code uses the primitive int everywhere, rather than the boxed integer. Let's follow that style.

Good idea. In the latest two years, I only wirte scala and give up Java entirely so that not sensitive to Java.

**Comment by shijinkui** [comment]
@StephanEwen review it again please.

**Comment by StephanEwen** [comment]
Thanks, picking this pull request up and merging it...

**Pull request closed without merging**